src/votekit/ballot_generator.py

```diff
@@ -1,70 +1,140 @@
+            (ex. {race: 0.5})
+            cohesion (dict): cohension factor for each bloc
+            alphas (dict): alpha for the dirchlet distribution of each bloc
+
+        Raises:
+            ValueError: if the voter proportion for blocs don't sum to 1
+            ValueError: slates and blocs are not the same
+
+        Returns:
+            BallotGenerator: initialized ballot generator
+        """
+
+        if sum(bloc_voter_prop.values()) != 1.0:
+            raise ValueError(
+                f"bloc proportions ({bloc_voter_prop.values()}) do not equal 1"
             )
-        else:
-            self.candidates = candidates
-        self.ballot_length = (
-            ballot_length if ballot_length is not None else len(self.candidates)
-        )
+        if slate_to_candidates.keys() != bloc_voter_prop.keys():
+            raise ValueError("slates and blocs are not the same")
 
-        self.parameterized = False
+        def _construct_preference_interval(
+            alphas: dict, cohesion: int, bloc: str, slate_to_cands: dict
+        ) -> dict:
+            intervals = {}
 
-        if hyperparams:
-            if isinstance(candidates, dict):  # add type error
-                self.set_params(candidates, **hyperparams)
-                self.parameterized = True
-            else:
-                raise TypeError(
-                    "'candidates' must be dictionary when hyperparameters are set"
-                )
-        else:
-            self.slate_to_candidate = {}
-            self.pref_interval_by_bloc: dict = {}
-            self.bloc_voter_prop: dict = {}
-            self.bloc_crossover_rate: dict = {}
+            for group, alpha in alphas.items():
+                num_cands = len(slate_to_cands[group])
+                probs = list(np.random.dirichlet([alpha] * num_cands))
+                for prob, cand in zip(probs, slate_to_cands[group]):
+                    if group == bloc:  # e.g W for W cands
+                        pi = cohesion
+                    else:  # e.g W for POC cands
+                        pi = 1 - cohesion
+                    intervals[cand] = pi * prob
+
+            return intervals
+
+        interval_by_bloc = {}
+        for bloc in bloc_voter_prop:
+            interval = _construct_preference_interval(
+                alphas[bloc], cohesion[bloc], bloc, slate_to_candidates
+            )
+            interval_by_bloc[bloc] = interval
+
+        if "candidates" not in data:
+            cands = list(
+                {cand for cands in slate_to_candidates.values() for cand in cands}
+            )
+            data["candidates"] = cands
+
+        if "pref_interval_by_bloc" not in data:
+            data["pref_interval_by_bloc"] = interval_by_bloc
+
+        if "bloc_voter_prop" not in data:
+            data["bloc_voter_prop"] = bloc_voter_prop
+
+        generator = cls(**data)
+
+        if isinstance(generator, AlternatingCrossover) or isinstance(
```

Comment:
tip: second argument to isinstance can take a tuple `isinstance(generator, (AlternatingCrossover, CambridgeSampler))`